File: mailbox/Mailbox.java


import java.util.*;


public class Mailbox {

	public static int max(int a,int b) {return a>b?a:b;}
	public static int min(int a,int b) {return a<b?a:b;}

	public static void main(String[] args) {
		Scanner scan =new Scanner(System.in);
		int numcases=scan.nextInt();
		int k=0,m=0;
		int[][][]table;
		for (int i=0;i<numcases;i++) {
			k=scan.nextInt();
			m=scan.nextInt();
			table=new int[k+1][m+1][m+1];
			initTable(table);
			System.out.println( optimalSolution(k,m,table));;
		}
	}
	public static void initTable(int[][][]table) {
		for (int i=0;i<table.length;i++) {
			for (int j=0;j<table[i].length;j++) {
				for (int k=0;k<table[i][j].length;k++) {
					table[i][j][k]=-1;
				}
			}
		}
	}


	public static int optimalSolution(int k, int m, int[][][]table) {
		return func(k,1,m,table);
	}

	public static int func(int k, int i, int j, int[][][]table) {
		if (i>j) {
			return 0;
		}
		if (table[k][i][j]>=0) {
			return table[k][i][j];
		}
		if (k ==1) {
			table[k][i][j]=j*(j+1)/2-i*(i-1)/2;
			return table[k][i][j];
		}
		int minval=Integer.MAX_VALUE;
		for (int s=i;s<=j;s++) {
			minval=min(minval,s+max(func(k-1,i,s-1,table),func(k,s+1,j,table)));
		}
		table[k][i][j]=minval;
		return minval;
	}

}

File: mailbox/mailbox.py

import sys


def create_array(*args):
	if not len(args):
		return -1
	A=[]
	args1=args[1:]
	for i in xrange(args[0]):
		A.append(create_array(*args1))
	return A

numcases=int(raw_input())


def func(k,i,j,table):
	if i>j:
		return 0
	if table[k][i][j]>=0:
		return table[k][i][j]
	if k ==1:
		table[k][i][j]=j*(j+1)/2-i*(i-1)/2
		return table[k][i][j]
	minval=sys.maxint
	for s in xrange(i,j+1):
		minval=min(minval,s+max(func(k-1,i,s-1,table),func(k,s+1,j,table)))
	table[k][i][j]=minval
	return minval



def optimal_sol(k,m,table):
	return func(k,1,m,table)
	#return min((func(k,m,t,table) for t in xrange(m,0,-1)))



for i in xrange(numcases):
	k,m=map(int,raw_input().strip().split())
	table= create_array(k+1,m+1,m+1)
	print optimal_sol(k,m,table)
